src/steadystateproblem.cpp

```diff
@@ -238,45 +243,71 @@ void SteadystateProblem::computeSteadyStateQuadrature(NewtonSolver *newtonSolver
+                                                 Model *model) {
+    /* Computes the integral over the adjoint state xB:
+     If the Jacobian has full rank, this has an anlytical solution, since
+     d/dt[ xB(t) ] = JB^T(x(t), p) xB(t) = JB^T(x_ss, p) xB(t)
+     This linear ODE system with time-constant matrix has the solution
+     xB(t) = exp( t * JB^T(x_ss, p) ) * xB(0)
+     This integral xQ over xB is given as the solution of
+     JB^T(x_ss, p) * xQ = xB(0)
+     So we first try to solve the linear system, if possible. */
+
+    /* copy content of xB into vector with integral */
+    xQ.copy(xB);
+
+    /* try to solve the linear system */
     try {
+        /* compute integral over xB and write to xQ */
         newtonSolver->prepareLinearSystemB(0, -1);
-        newtonSolver->solveLinearSystem(xB);
+        newtonSolver->solveLinearSystem(xQ);
+        /* Compute the quadrature as the inner product xQ * dxotdp */
+        getQBfromQ(model, xQ, xQB);
+        /* set flag that quadratures is available (for processing in rdata) */
+        hasQuadrature_ = true;
     } catch (NewtonFailure const &ex) {
-        if (ex.error_code == AMICI_SINGULAR_JACOBIAN)
-            throw NewtonFailure(ex.error_code, "Steady state backward "
-                                "computation failed due to unsuccessful "
-                                "factorization of RHS Jacobian.");
-        throw NewtonFailure(ex.error_code, "Steady state backward "
-                            "computation failed.");
+        hasQuadrature_ = false;
     }
+}
 
-    /* Compute the quadrature as the inner product xBI * dxotdp */
-    if (model->pythonGenerated) {
-        /* fill dxdotdp with current values */
-        const auto& plist = model->getParameterList();
-        model->fdxdotdp(t, x, x);
+void SteadystateProblem::getQuadratureBySimulation(const Solver *solver,
+                                                   Model *model) {
+    /* If the Jacobian is singular, the integral over xB must be computed
+       by usual integration over time, but  simplifications can be applied:
+       x is not time dependent, no forward trajectory is needed. */
 
-        if (model->ndxdotdp_explicit > 0)
-            model->dxdotdp_explicit.multiply(xQB.getNVector(),
-                                             xB.getNVector(), plist, true);
-        if (model->ndxdotdp_implicit > 0)
-            model->dxdotdp_implicit.multiply(xQB.getNVector(),
-                                             xB.getNVector(), plist, true);
-    } else {
-        for (int ip=0; ip<model->nplist(); ++ip)
-            xQB[ip] = N_VDotProd(xB.getNVector(),
-                                 model->dxdotdp.getNVector(ip));
+    /* set starting timepoint for the simulation solver */
+    t = 0;
+    /* xQ was written in getQuadratureByLinSolve() -> reset */
+    xQ.reset();
+    /* initialize the Jacobian */
+    model->writeSteadystateJB(t, 0, x, x, xB, xB, xB);
```

Comment:
Here's where we reset the Model Jacobian... As said, not sure whether we maybe want to avoid it and go another way.

Reply:
Hm... May reduce the overall dependencies a bit, yes. However, atm this only works with KLU. In the near future, `writeSteadystateJB()` might include a switch depending on the linear solver... Not 100% sure yet, what's the best solution...
For the moment however, we can go with your solution, as it makes things look a little cleaner... and avoids exposing this function via abstract_model to the outside world, which is a good thing.